data_structures/binary_search_tree2.cpp

```diff
@@ -0,0 +1,539 @@
+    assert(!fail_res);
+    assert(tree.size() == 0);
+
+    std::cout << "ok" << std::endl;
+}
+
+/**
+ * @brief Function for testing contains().
+ */
+void test_contains() {
+    std::cout << "Testing BST contains...";
+
+    binary_search_tree<int> tree;
+    tree.insert(5);
+    tree.insert(4);
+    tree.insert(3);
+    tree.insert(6);
+
+    assert(tree.contains(5));
+    assert(tree.contains(4));
+    assert(tree.contains(3));
+    assert(tree.contains(6));
+    assert(!tree.contains(999));
+
+    std::cout << "ok" << std::endl;
+}
+
+/**
+ * @brief Function for testing find_min().
+ */
+void test_find_min() {
+    std::cout << "Testing BST find_min...";
+
+    int min = 0;
+    binary_search_tree<int> tree;
+    assert(!tree.find_min(min));
+
+    tree.insert(5);
+    tree.insert(4);
+    tree.insert(3);
+    tree.insert(6);
+
+    assert(tree.find_min(min));
+    assert(min == 3);
+
+    std::cout << "ok" << std::endl;
+}
+
+/**
+ * @brief Function for testing find_max().
+ */
+void test_find_max() {
+    std::cout << "Testing BST find_max...";
+
+    int max = 0;
+    binary_search_tree<int> tree;
+    assert(!tree.find_max(max));
+
+    tree.insert(5);
+    tree.insert(4);
+    tree.insert(3);
+    tree.insert(6);
+
+    assert(tree.find_max(max));
+    assert(max == 6);
+
+    std::cout << "ok" << std::endl;
+}
+
+/**
+ * @brief Function for testing get_elements_inorder().
+ */
+void test_get_elements_inorder() {
+    std::cout << "Testing BST get_elements_inorder...";
+
+    binary_search_tree<int> tree;
+    tree.insert(5);
+    tree.insert(4);
+    tree.insert(3);
+    tree.insert(6);
+
+    std::vector<int> expected = {3, 4, 5, 6};
+    std::vector<int> actual = tree.get_elements_inorder();
+    assert(actual == expected);
+
+    std::cout << "ok" << std::endl;
+}
+
+/**
+ * @brief Function for testing get_elements_preorder().
+ */
+void test_get_elements_preorder() {
+    std::cout << "Testing BST get_elements_preorder...";
+
+    binary_search_tree<int> tree;
+    tree.insert(5);
+    tree.insert(4);
+    tree.insert(3);
+    tree.insert(6);
+
+    std::vector<int> expected = {5, 4, 3, 6};
+    std::vector<int> actual = tree.get_elements_preorder();
+    assert(actual == expected);
+
+    std::cout << "ok" << std::endl;
+}
+
+/**
+ * @brief Function for testing get_elements_postorder().
+ */
+void test_get_elements_postorder() {
+    std::cout << "Testing BST get_elements_postorder...";
+
+    binary_search_tree<int> tree;
+    tree.insert(5);
+    tree.insert(4);
+    tree.insert(3);
+    tree.insert(6);
+
+    std::vector<int> expected = {3, 4, 6, 5};
+    std::vector<int> actual = tree.get_elements_postorder();
+    assert(actual == expected);
+
+    std::cout << "ok" << std::endl;
+}
+
+int main() {
+    test_insert();
+    test_remove();
+    test_contains();
+    test_find_max();
+    test_find_min();
+    test_get_elements_inorder();
+    test_get_elements_preorder();
+    test_get_elements_postorder();
+}
```

Comment:
```suggestion
}

```
Please leave a new-line character at the EOF